### PR TITLE
IF-284: Asset wizard thresholds and save asset

### PR DESCRIPTION
**Pull request by jsylupp**

### CODE REVIEW

**Comment by mattmikulina** [comment]
Task linked: [IF-284 Add Thresholds to metrics (page 3) and save asset](https://app.clickup.com/t/2177764/IF-284)

File: fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard-step/asset-wizard-step-metrics-thresholds/asset-wizard-step-metrics-thresholds.component.ts [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    // As demoinsert data also have thresholds for non numeric data, force to remove them before saving
+    if (quantityDataType !== QuantityDataType.NUMERIC) {
+      return null;
+    }
+
+    const lowerValue = thresholdGroup.get(type + 'Lower').value;
+    const upperValue = thresholdGroup.get(type + 'Upper').value;
+    const valueExist: boolean = lowerValue || upperValue;
+    if (!valueExist) {
+      return null;
```
Comment (seemueller) [inline fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard-step/asset-wizard-step-metrics-thresholds/asset-wizard-step-metrics-thresholds.component.ts:69]: may not use many returns in one method

File: fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard.component.ts [hunk 2]
```diff
@@ -95,6 +110,61 @@ export class AssetWizardComponent implements OnInit {
...
+/*  onUpdateAssetSeries() {
+    if (this.assetSeries.id) {
+      this.assetSeriesService.editItem(this.assetSeries.id, this.assetSeries)
+        .subscribe(newAssetSeries => this.assetSeries = newAssetSeries);
+    } else {
+      this.assetSeries.companyId = this.companyId;
+      this.assetSeriesService.createItem(this.assetSeries.companyId, this.assetSeries.assetTypeTemplateId)
+        .subscribe(newAssetSeries => this.assetSeries = newAssetSeries);
+    }
+  }*/
```
Comment (seemueller) [inline fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard.component.ts:166]: remove unused code

File: fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard.component.ts [hunk 3]
```diff
@@ -95,6 +110,61 @@ export class AssetWizardComponent implements OnInit {
...
+      this.asset.guid = this.assetForm.get('guid')?.value;
+      this.asset.serialNumber = this.assetForm.get('serialNumber')?.value;
+      this.asset.constructionDate = this.assetForm.get('constructionDate')?.value;
+      this.asset.installationDate = this.assetForm.get('installationDate')?.value;
+      this.asset.protectionClass = this.assetForm.get('protectionClass').value;
+      this.asset.handbookKey = this.assetForm.get('handbookKey').value;
+      this.asset.videoKey = this.assetForm.get('videoKey')?.value;
+      this.asset.imageKey = this.assetForm.get('imageKey')?.value;
+
+      console.log('asset to save:', this.asset);
```
Comment (seemueller) [inline fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard.component.ts:147]: remove log

File: fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard.component.ts [hunk 4]
```diff
@@ -119,8 +189,10 @@ export class AssetWizardComponent implements OnInit {
     this.assetSeriesResolver.resolve(this.activatedRoute.snapshot);
     this.assetResolver.resolve(this.activatedRoute.snapshot);
     this.assetTypesResolver.resolve().subscribe();
-    this.isLoading$ = this.assetSeriesQuery.selectLoading();
+    this.fieldsResolver.resolve().subscribe();
+    this.isLoading$ = this.assetSeriesQuery.selectLoading(); // TODO: is this correct?
```
Comment (seemueller) [inline fusionfrontend/src/app/fleet/components/content/asset-wizard/asset-wizard.component.ts:193]: open todo here?

File: fusionfrontend/src/assets/sass/components/_labels.scss [hunk 5]
```diff
@@ -15,6 +15,12 @@
   opacity: 1;
 }
 
+// TODO: good naming?
```
Comment (seemueller) [inline fusionfrontend/src/assets/sass/components/_labels.scss:18]: yes ;)

File: fusionfrontend/src/assets/sass/components/_primeaccordion.scss [hunk 6]
```diff
@@ -0,0 +1,40 @@
+@import 'base/typography';
```
Comment (seemueller) [inline fusionfrontend/src/assets/sass/components/_primeaccordion.scss:1]: missing licence comment?

File: fusionbackend/src/main/java/io/fusion/fusionbackend/rest/fleetmanager/FleetAssetRestService.java [hunk 7]
```diff
@@ -68,11 +76,74 @@ public AssetDto updateAsset(@PathVariable final Long companyId,
...
-                false);
+    public AssetDto createAsset(@PathVariable final Long companyId,
+                                @PathVariable final Long assetSeriesId,
+                                @RequestBody final AssetDto assetDto) {
+        return assetMapper.toDto(
+                assetService.createAssetAggregate(companyId, assetSeriesId, assetMapper.toEntity(assetDto)),
+                true);
+    }
+
+    // TODO: this should probably not be allowed
```
Comment (seemueller) [inline fusionbackend/src/main/java/io/fusion/fusionbackend/rest/fleetmanager/FleetAssetRestService.java:87]: open todo?

File: fusionbackend/src/main/java/io/fusion/fusionbackend/rest/fleetmanager/FleetAssetRestService.java [hunk 8]
```diff
@@ -68,11 +76,74 @@ public AssetDto updateAsset(@PathVariable final Long companyId,
...
+    @PostMapping(path = "/companies/{companyId}/assets/{assetId}/fieldinstances")
+    public FieldInstanceDto createFieldInstance(@PathVariable final Long companyId,
+                                                @PathVariable final Long assetId,
+                                                @RequestBody final FieldInstanceDto fieldInstanceDto) {
+        return fieldInstanceMapper.toDto(
+                assetService.createFieldInstance(companyId, assetId,
+                        fieldInstanceMapper.toEntity(fieldInstanceDto)), false);
+    }
+
+    // TODO: what values should be allowed to be changed
```
Comment (seemueller) [inline fusionbackend/src/main/java/io/fusion/fusionbackend/rest/fleetmanager/FleetAssetRestService.java:97]: open todo?

File: fusionbackend/src/main/java/io/fusion/fusionbackend/rest/fleetmanager/FleetAssetRestService.java [hunk 9]
```diff
@@ -68,11 +76,74 @@ public AssetDto updateAsset(@PathVariable final Long companyId,
...
+    @PatchMapping(path = "/companies/{companyId}/assets/{assetId}/fieldinstances/{fieldInstanceId}")
+    public FieldInstanceDto updateFieldInstance(@PathVariable final Long companyId,
+                                                @PathVariable final Long assetId,
+                                                @PathVariable final Long fieldInstanceId,
+                                                @RequestBody final FieldInstanceDto fieldInstanceDto) {
+        return fieldInstanceMapper.toDto(assetService.updateFieldInstance(companyId, assetId,
+                fieldInstanceId, fieldInstanceMapper.toEntity(fieldInstanceDto)), false);
+    }
+
+    // TODO: this should probably not be allowed
```
Comment (seemueller) [inline fusionbackend/src/main/java/io/fusion/fusionbackend/rest/fleetmanager/FleetAssetRestService.java:107]: open todo?

File: fusionbackend/src/main/java/io/fusion/fusionbackend/service/FieldInstanceService.java [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        if (fieldInstance.getCriticalThreshold() == null) {
+            return 0;
+        }
+        return ((fieldInstance.getCriticalThreshold().getValueUpper() != null) ? 1 : 0)
+                + ((fieldInstance.getCriticalThreshold().getValueLower() != null) ? 1 : 0);
+    }
+
+    private boolean existPairwiseValues(final int absoluteValuesCount,
+                                        final int idealValuesCount,
+                                        final int criticalValuesCount) {
```
Comment (seemueller) [inline fusionbackend/src/main/java/io/fusion/fusionbackend/service/FieldInstanceService.java:118]: here ist maby a shortcut with the XOR operator possible

!((a != null) ^ (b != null))

**Review by seemueller — CHANGES_REQUESTED** [review]

**Pull request merged**